### ISSUE TITLE
integrate with GitHub for Requests project management

### ISSUE THREAD

**Issue by josh-chamberlain** [Police-Data-Accessibility-Project/data-sources-app#299]
### Context

We'll use github to communicate about requests. This integration allows us to create issues, and read their status from the app.

- [here's the model project](https://github.com/orgs/Police-Data-Accessibility-Project/projects/26).
  - it has some issues in it, all manually-generated. Some of them won't have all the metadata attached, but the project-level custom fields are the target.
- requires #256 
- requires https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/74
- related to https://github.com/Police-Data-Accessibility-Project/data-projects/issues/5

### Requirements

- #74 
- [x] #464
- [x] #465
- [x] when a request hits the `ready to start` status, create a new GitHub issue in the project above, in the [data-projects repo](https://github.com/Police-Data-Accessibility-Project/data-projects)
- [x] Modify the project automations (i.e., in https://github.com/orgs/Police-Data-Accessibility-Project/projects/26/workflows ) so that, when an issue is added to the repository, it is automatically added to the project and given a specific status
- [x] GitHub is then the source of truth for `request_status`; infer this from GitHub by checking for it periodically
- [x] when the page is loaded, limited to once every minute—this may result in fewer refreshes overall)
- [x] `record_types_required` can be labels.
   - we should sync these back to our database too, for portability
- [x] add an issue template to the repo that tells people they should not make a request by making an issue, and points to the intake form
- [x] Set up the following environment variables in Digital Ocean:
  - [x] GH_API_ACCESS_TOKEN: Used to authenticate with GitHub when adding or getting information about Github Issues. Must correspond to the `access_token` provided by GitHub. 
  - [x] GH_ISSUE_REPO_OWNER: Identifiers the github user. In this case, `Police-Data-Accessibility-Project`
  - [x] GH_ISSUE_REPO_NAME: Identifies the repository, in this case, `data-requests`, where the issue will be created, and where information about issues will be retrieved. Can be customized for local development.
- [x] Create Notion Documentation describing how everything works under the hood
- [x] Ensure that issues appear in the following format:
```markdown
## Submission notes
{submission notes}

## Location
{location(s) covered}

## Data requirements
{data requirements}
```

### @josh-chamberlain 
- [x] Determine repository where issues will be published. 
  - https://github.com/Police-Data-Accessibility-Project/data-projects
- [x] Create Github Access Token and send to @maxachis. See specific permissions required [here](https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/299#issuecomment-2405951278).

### @maxachis 
- [x] Set up envars in Stage
- [x] Push changes into Stage
- [x] Run Prod-To-Stage Migration to bring test data request into stage environment
- [x] Run `/synchronize` and see that a new issue is created from the test data request 
- [x] Update the status of the test data request and run synchronize again to see that it proliferates into the dev database. 
- [x] Once all that is successful, add the envars to Prod
- [x] Merge latest changes into main
- [x] Create user `github_sync@pdap.io` and assign permission `github_sync`
- [x] Set up automation job that will periodically call `/synchronize` in the prod environment, using the above user
- [x] Run `/synchronize` again to confirm its functionality
- [x] Do the final merge into main.

### Frontend
- [ ] https://github.com/Police-Data-Accessibility-Project/pdap.io/issues/267
- [ ] https://github.com/Police-Data-Accessibility-Project/pdap.io/issues/268

#### The GitHub Discussions option

- we _could_ use github discussions to prevent new requests from being created except through the app, but this seems complicated
- discussions are kind of higher level research questions, and issues are tasks within them. discussions don't have project management. if we use discussions, we need an orderly way to extract the requests from them—especially for simple cases where 1 request → 1 discussion → 1 task → response, we want to cut the noise a bit.

**Comment by maxachis** [comment]
>     * maybe they go into a private repo at first, and get moved into a public repo when they're ready for volunteers

One option is to utilize [GitHub discussions](https://docs.github.com/en/discussions/quickstart), which would enable us to review and discuss data requests before converting them to an issue (which can be easily done using Discussions functionality). GitHub discussions, as I understand it, is tied to the visibility of the repository it is connected with. 

Still, having them exist as discussions allows us to refine and potentially talk more with the request provider before converting it to a distinct issue, and avoids having to maintain two separate repositories. We can set it so that users with read access [can be disallowed ](https://docs.github.com/en/organizations/managing-organization-settings/managing-discussion-creation-for-repositories-in-your-organization)from creating discussions. 

At the very least, I think it's worth playing around with to see whether it can work for our purposes.

**Comment by maxachis** [comment]
@josh-chamberlain Here is the MVP workflow as I am envisioning it (from the backend's perspective). We can expand on this, but I want to make sure I've got the core functionality down.

1. The `POST` endpoint for `/data-requests` is called, creating the data request in the backend
2. At some later point, the `POST` endpoint for `/data-requests/github-issue/{ID}` is called, which takes the data request as it exists in the database and creates an issue in the Data Projects repository using that data. The `github_issue_url` field is updated in the database, and possibly the `request_status`.

Does this align with what you're thinking?

**Comment by josh-chamberlain** [comment]
@maxachis sorry, I never responded! Yes, this is what I'm thinking. I made some changes to the issue yesterday. We should wait to create the issue until the status hits that point.

**Comment by maxachis** [comment]
# Github Rest API

https://docs.github.com/en/rest?apiVersion=2022-11-28

**Comment by maxachis** [comment]
@josh-chamberlain Alright, after doing some snurfling, here's what I've determined as far as the implementation goes.

I'll also re: @joshuagraber here, so he can sanity check my thinking on this. 

## Implementation
This feature, as currently designed would involve multiple components:
1. Modifying the project automations (i.e., in https://github.com/orgs/Police-Data-Accessibility-Project/projects/26/workflows ) so that, when an issue is added to the repository, it is automatically added to the project and given a specific status
2. Setting up a `/github/data-requests/issues` endpoint, whose `POST` functionality adds an issue to GitHub with the given information,
   1. This requires a GraphQL call
3. Setting up a `/github/data-requests/synchronize` endpoint, whose `POST` functionality
   1. Uses a GraphQL call to get all issues (and their statuses) in the repository
   2. Checks to see if they don't exist in the Database, and adds them to the database if not
   3. Checks to see if their status is the same as it is in the database, and updates the database status if not
4. Creating an automation manager job which calls `/github/data/requests/synchronize` on a periodic basis

## This seems convoluted: Is there any other option?

Not as far as I can tell if we want to use GitHub projects, I'm afraid. There [are no GitHub Action triggers](https://github.com/orgs/community/discussions/40848) (https://github.com/orgs/community/discussions/40848) for when a Project's status is changed, nor any Webhooks that do the same, so our only option is to repeatedly poll for updates and modify them in the database as they occur.

However, if we were instead interested in using [the future Retool ](https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/74)option to manage data requests, that would (in theory) provide us with a means for monitoring and manipulating data request statuses without having to use the clunky synchronization option. However, if we did that, we would lose the tight GitHub integration.

**Comment by josh-chamberlain** [comment]
@maxachis I don't think we should add `issues` from github as PDAP requests—syncing status makes sense, but creating requests from our app should be the one and only workflow (at least for now) for simplicity and clarity. We'd only be polling GitHub for status on occasion; we could do this _on page view_ and/or _periodically_.

**Comment by maxachis** [comment]
> @maxachis I don't think we should add `issues` from github as PDAP requests—syncing status makes sense, but creating requests from our app should be the one and only workflow (at least for now) for simplicity and clarity. We'd only be polling GitHub for status on occasion; we could do this _on page view_ and/or _periodically_.

@josh-chamberlain Got it! I've updated the comment accordingly!

**Comment by maxachis** [comment]
We'll need a classic API token (not a personal access token) to work this: Here are the permissions required:

![image](https://github.com/user-attachments/assets/43dc8fb1-eb82-4d6b-9c27-46c7d0d6170a)
![image](https://github.com/user-attachments/assets/01effcc8-2205-4a60-ba4d-3a675017c95e)
![image](https://github.com/user-attachments/assets/90ae069d-4ce0-45e7-a88b-d246a5ac7ee6)

**Comment by josh-chamberlain** [comment]
@maxachis hmm, is there a way to maybe use a different (private) github project that would allow us to do end to end testing?

**Comment by maxachis** [comment]
> @maxachis hmm, is there a way to maybe use a different (private) github project that would allow us to do end to end testing?

Oh, absolutely -- indeed, I tested components with my own personal repo -- just not all of the components combined. The nature of what we're doing with Github would likely render the parts where we interact with Github directly manual (unless we want to automate test interactions with Github as well), but it can all be tested.

**Comment by josh-chamberlain** [comment]
@maxachis are you able to share screenshots or other evidence of what the issues look like upon creation? I am a bit nervous to leave it until launch to find out I'm picky about the format or something.

**Comment by maxachis** [comment]
Certainly! Here's a (very coarse) example of one:

![image](https://github.com/user-attachments/assets/4b150481-c704-442f-9e2d-1333f1a5a0d4)

The title and submission notes are randomly generated garble. Previously, this included only the first 50 characters of submission notes, but with title we can now have the title be the, well, title.

To wireframe it somewhat, the format of the body is:

```
Submission Notes: {whatever the submission notes are}

Data Requirements:
{whatever the Data Requirements are}
```

But this is a very rough draft and can easily be updated.

**Comment by maxachis** [comment]
@josh-chamberlain There's a few bits of thread we probably want to take care of:

1. Currently, all data requests in the database are listed as `intake` -- this is a product of the fact that data requests did not previously have a status. Technically, this shouldn't stop us from setting up the connective tissue, but we probably should resolve that to avoid confusion.
2. To do the above, we may need to go in and manually update the status of the data requests, which may in turn be contingent on fixing up the retool for data requests.
3. We'll probably want to set up a test repository for publishing data requests as we test this in dev, and I'll need to make sure the internal logic is robust to enable that.

**Comment by maxachis** [comment]
Related to the above, this issue is blocked until the completion of #643

**Comment by josh-chamberlain** [comment]
@maxachis yes, I had planned to address these in #580...is this a chicken and egg problem? The issues that were manually created for the requests (linked in the requests table) have a `status` in the project: https://github.com/orgs/Police-Data-Accessibility-Project/projects/26

**Comment by maxachis** [comment]
As far as documenting how this process works, I'll discuss this in the notion page [here](https://www.notion.so/pdap/Github-Synchronization-1b080fce1fec80208365c158dee2377a?pvs=4).

**Comment by maxachis** [comment]
In terms of synchronization, I'm thinking of making a few adjustments to it

## Remove `/github/data-requests/issues` `POST` and implement that logic into `/synchronize`

Rather than have an endpoint that requires us to manually add each data request to GitHub, we would instead have it that synchronization will
1. Look at any data request that is marked as `Ready to Start` and doesn't have an associated GitHub Issue URL
2. Add that as an issue to the GitHub repository and provide the GitHub Issue URL.

**Comment by maxachis** [comment]
> [@maxachis](https://github.com/maxachis) yes, I had planned to address these in [#580](https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/580)...is this a chicken and egg problem? The issues that were manually created for the requests (linked in the requests table) have a `status` in the project: https://github.com/orgs/Police-Data-Accessibility-Project/projects/26
> 
> actually, I am now seeing that github_issue_url did not make it into the database. It's in Airtable: https://airtable.com/app473MWXVJVaD7Es/tblV2ftkAwaNJfDTJ/viwQk3dEEJyX3zGlQ?blocks=bipb7pe2uMBSXKRwQ
> 
> do we need to do some kind of one-time sync?

@josh-chamberlain We could do a one-time sync, or we could manually add them into the database via a migration job! I'd prefer the latter just because it's easier to do the drudgery of adding GitHub issues en masse once than creating a bespoke migration script.

**Comment by maxachis** [comment]
> [@maxachis](https://github.com/maxachis) yes, I had planned to address these in [#580](https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/580)...is this a chicken and egg problem? The issues that were manually created for the requests (linked in the requests table) have a `status` in the project: https://github.com/orgs/Police-Data-Accessibility-Project/projects/26
> 
> actually, I am now seeing that github_issue_url did not make it into the database. It's in Airtable: https://airtable.com/app473MWXVJVaD7Es/tblV2ftkAwaNJfDTJ/viwQk3dEEJyX3zGlQ?blocks=bipb7pe2uMBSXKRwQ
> 
> do we need to do some kind of one-time sync?

@josh-chamberlain This can be addressed with a migration script, which I can work on setting up using this CSV, taken from Airtable:

[Data Requests-PDAP Data Requests.csv](https://github.com/user-attachments/files/19184301/Data.Requests-PDAP.Data.Requests.csv)

**Comment by maxachis** [comment]
@josh-chamberlain Well, the first part of the synchronization is a success! An issue was created in data-requests and the GitHub issue URL reflected in the database. 

However, there are a few snags:

1. All the existing data requests exist in either the `data-projects` or the `scrapers` repository. If we can move those into `data-requests`, that will simplify the logic. 
2. The issue added was not automatically added to a project nor given a `Ready to start` label -- we'll need to set up the logic so that this is done automatically for any issue created.

**Comment by josh-chamberlain** [comment]
@maxachis nice! 

1. sure, 3 issues transferred:
  - https://github.com/Police-Data-Accessibility-Project/data-requests/issues/96
  - https://github.com/Police-Data-Accessibility-Project/data-requests/issues/97
  - https://github.com/Police-Data-Accessibility-Project/data-requests/issues/98
2. to be clear, by "automatically", you mean _after it was processed in Retool?_ The idea is that the request exists outside GitHub, then when the approver selects "ready to start", the issue is created. I think we are on the same page because , just checkin.

**Comment by maxachis** [comment]
> to be clear, by "automatically", you mean after it was processed in Retool? The idea is that the request exists outside GitHub, then when the approver selects "ready to start", the issue is created. I think we are on the same page because , just checkin.

@josh-chamberlain Correct! What I'm talking about is that there is logic in GitHub that can automatically assign new issues to a project and give it a status. This is detailed in [this GitHub documentation](https://docs.github.com/en/issues/planning-and-tracking-with-projects/automating-your-project/adding-items-automatically).

Basically, once the script adds an issue, GitHub takes over and adds it to the project and gives it a status of `Ready to start`.

**Comment by maxachis** [comment]
@josh-chamberlain I've got the synchronization set up! I'm now working on the `record_types_required` component.

But a quick question: 

> `work_required` is converted to labels on the issue

What is `work_required` in this case? I don't see this in the database.

**Comment by josh-chamberlain** [comment]
@maxachis `work_required` is outdated! `record types` as labels sound reasonable enough—but then we'd want to sync them both ways...it's starting to sound a little less reasonable.

**Comment by maxachis** [comment]
@josh-chamberlain I have now made the following changes:
- Added labels for each record type to the `data-requests` repository. These have the `RT-` prefix for easy identifiability
- Updated synchronization such that record types included on an issue not yet added to GitHub are included in the GitHub issue as labels on creation
- Updated synchronization such that any modifications to labels in the issue are propagated to the database

Notably, I have not done the reverse -- if a data request that already has an associated GitHub issue has its `record_types_required` overwritten, that will not persist during the next synchronization -- they will be overwritten by the labels on the GitHub issue. 

That's for two reasons:
1. I figure this aligns with the whole idea of GitHub being a source of truth
2. It's a PITA to create two-way logic that updates based on whichever was most recently updated. Possibly so much as to be impractical to attempt.

**Comment by maxachis** [comment]
Notion documentation created! 

https://www.notion.so/pdap/GitHub-Requests-Synchronization-1b080fce1fec80208365c158dee2377a

And with that, I think we can finally close this issue! Do you agree, @josh-chamberlain ?
